src/box/txn_limbo.c

```diff
@@ -240,19 +250,12 @@ txn_limbo_assign_local_lsn(struct txn_limbo *limbo,
 	assert(lsn > 0);
 
 	entry->lsn = lsn;
-	/*
-	 * The entry just got its LSN after a WAL write. It could
-	 * happen that this LSN was already ACKed by some
-	 * replicas. Update the ACK counter to take them into
-	 * account.
-	 */
-	struct vclock_iterator iter;
-	vclock_iterator_init(&iter, &limbo->vclock);
-	int ack_count = 0;
-	vclock_foreach(&iter, vc)
-		ack_count += vc.lsn >= lsn;
-	assert(ack_count >= entry->ack_count);
-	entry->ack_count = ack_count;
+	if (entry == limbo->entry_to_confirm) {
+		limbo->ack_count = vclock_count_ge(
+			&limbo->vclock, entry->lsn);
+		if (limbo->ack_count >= replication_synchro_quorum)
+			txn_limbo_confirm(limbo);
```

Comment:
I see the following segfaults when running the tests:
```
[060] master | #1  0x557dea5a48c5 in crash_collect+256
[060] master | #2  0x557dea5a5301 in crash_signal_cb+100
[060] master | #3  0x76d759e42520 in __sigaction+80
[060] master | #4  0x557dea5ae8d5 in cord_reset_slice+263
[060] master | #5  0x557dea5afdba in fiber_yield_impl+262
[060] master | #6  0x557dea5aff26 in fiber_yield+18
[060] master | #7  0x557dea465a77 in journal_write+40
[060] master | #8  0x557dea466117 in journal_write_row+188
[060] master | #9  0x557dea4107a5 in synchro_request_write+85
[060] master | #10 0x557dea4108fb in txn_limbo_write_synchro+190
[060] master | #11 0x557dea4109e8 in txn_limbo_write_confirm+214
[060] master | #12 0x557dea411552 in txn_limbo_confirm+622
[060] master | #13 0x557dea4100cd in txn_limbo_assign_local_lsn+337
[060] master | #14 0x557dea410113 in txn_limbo_assign_lsn+67
[060] master | #15 0x557dea40a3e2 in txn_on_journal_write+1616
[060] master | #16 0x557dea473ac4 in journal_async_complete+102
[060] master | #17 0x557dea474532 in tx_schedule_queue+54
[060] master | #18 0x557dea4746be in tx_complete_batch+133
[060] master | #19 0x557dea5bad9e in cmsg_deliver+52
[060] master | #20 0x557dea5bb72e in cbus_process+100
[060] master | #21 0x557dea431cf3 in tx_prio_cb(ev_loop*, ev_watcher*, int)+185
[060] master | #22 0x557dea979528 in ev_invoke_pending+205
[060] master | #23 0x557dea97a4ce in ev_run+1529

```

This must be because you started writing CONFIRM from the same fiber which processes a WAL write confirmation (sched).

Previously we explicitly didn't ACK such transactions right on their WAL write confirmation to avoid this. Take a look at txn.c, for example:
https://github.com/tarantool/tarantool/blob/ea6062964828015e5e1f376c6ec2f0b7829b40c2/src/box/txn.c#L1221-L1229
As you can see, instead of ack-ing the transaction from sched right on its WAL write, we ack it from a fiber which owns the transaction.

Reply:
You can cooperate with @CuriousGeorgiy, he apparently might need to solve the same problem for async commit of sync txns.

Reply:
Working on this in #10626.

Reply:
```diff
void
txn_limbo_assign_local_lsn(struct txn_limbo *limbo,
			   struct txn_limbo_entry *entry, int64_t lsn)
{
	assert(limbo->owner_id != REPLICA_ID_NIL);
	assert(limbo->owner_id == instance_id);
	assert(entry->lsn == -1);
	assert(lsn > 0);

	entry->lsn = lsn;

-	if (entry == limbo->entry_to_confirm) {
-		limbo->ack_count = vclock_count_ge(
-			&limbo->vclock, entry->lsn);
-		if (limbo->ack_count >= replication_synchro_quorum)
-			txn_limbo_confirm(limbo);
-	}
+	if (entry == limbo->entry_to_confirm)
+		limbo->ack_count = vclock_count_ge(&limbo->vclock, entry->lsn);
}
```
I fixed it just like that. Since before this patch the behavior was essentially the same - `CONFIRM` will only be written if the limbo owner said ACK, even if enough replicas (>= `replication_synchro_quorum`) have confirmed this transaction.